### PR TITLE
networkmanager: Checkpoints

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
Early demo: https://www.youtube.com/watch?v=cVtKw8ZAhQ8

**Comment by mvollmer** [comment]
To do:
 - tests
 - wrap slave adding/removing/onoffing in checkpoints, too.

**Comment by mvollmer** [comment]
Rolling back a checkpoint does not remove devices that were created after the checkpoint.

**Comment by mvollmer** [comment]
Clicking "Make these changes anway" does nothing, probably because the internal model has already been set back to the pre-checkpoint values.  The freeze/apply approach needs to be changed, I guess.

**Comment by mvollmer** [comment]
This needs to work also with NM versions that don't have checkpoints, of course.

**Comment by mvollmer** [comment]
> This needs to work also with NM versions that don't have checkpoints, of course.

Done.

**Comment by mvollmer** [comment]
> Clicking "Make these changes anway" does nothing, probably because the internal model has already been set back to the pre-checkpoint values. The freeze/apply approach needs to be changed, I guess.

Done.

**Comment by mvollmer** [comment]
> wrap slave adding/removing/onoffing in checkpoints, too.

Done.

**Comment by mvollmer** [comment]
> tests

Added one.

**Comment by mvollmer** [comment]
> Rolling back a checkpoint does not remove devices that were created after the checkpoint.

https://bugzilla.redhat.com/show_bug.cgi?id=1378393

**Comment by andreasn** [comment]
```
Switching off ens3 has broken the connection to the machine. It has been switched back on.
```

The wording needs a tiny little bit of tweaking.
```
Switching off ens3 broke the connection to the system. It was switched back on.
```

**Comment by andreasn** [comment]
I wonder if "ens3" (or whatever the connection in question is called) need to be identified a bit more, so in that case it would be "The interface ens3..."

**Comment by mvollmer** [comment]
> The wording needs a tiny little bit of tweaking.

Done.

**Comment by mvollmer** [comment]
> I wonder if "ens3" (or whatever the connection in question is called) need to be identified a bit more, so in that case it would be "Switching off the interface **ens3**..."

Done.

**Comment by dperpeet** [comment]
needs a rebase

**Comment by mvollmer** [comment]
> needs a rebase

Done.

**Comment by andreasn** [comment]
I did some quick, 5 minute usability tests with three participants. Asked them to log in to 192.168.1.218 and turn off the interface for 192.168.1.218.
These are the results: http://pastebin.com/kKTWWEaM
It is very unscientific, but gives some pointers about some current issues.

**Comment by andreasn** [comment]
As far as I can see, the biggest issues are lack of feedback, once the ON/OFF button is flipped, and unclear information exactly what happened, what exactly it was that happened, what got disconnected from what.

**Comment by stefwalter** [comment]
Nice research @andreasn

**Comment by andreasn** [comment]
So looking at the results, here is one way to solve it:
mockup

This gives clear feedback that something is happening, rather than just freezing up, and gives a clear option between keeping the connection (recommended), or shutting the connection off.
I'm not 100% sure about the timing of the full-page connection-tester, but something needs to happen immediately, just like you get instant feedback when you shut off a connection you're not sitting on.

**Comment by andreasn** [comment]
Another option could be to **always** trigger a spinner inside the network pane, even when it will succeed.
mockup

**Comment by mvollmer** [comment]
@andreasn, thanks for the research!

When making changes via a dialog, we can just keep that dialog open with a spinner and "Testing connection" text until the change either succeeds or has been rolled back.  I'll do that, okay?

**Comment by andreasn** [comment]
> When making changes via a dialog, we can just keep that dialog open with a spinner and "Testing connection" text until the change either succeeds or has been rolled back.

At the point you get the dialog, you already lost your connection once. "Keep connection" in this case does nothing, it just closes the dialog (like the Close button does right now). The switch off button would switch the thing off.

**Comment by andreasn** [comment]
The dialog only comes up once you lost the connection

**Comment by mvollmer** [comment]
>> When making changes via a dialog, we can just keep that dialog open with a spinner and "Testing connection" text until the change either succeeds or has been rolled back.

> At the point you get the dialog, you already lost your connection once. 

Sorry, I wasn't clear enough.  I was not talking about the dialog from your mockup (that informs the user about losing the connection), but dialogs such as the "IP Settings" dialog where you can change the IP address.

The flow with those dialogs could be either:

      "IP Settings" open -> make change -> "Apply" -> dialog closes -> big "testing connection" spinner appears -> "Connection will be lost" dialog opens.

Or:

      "IP Settings" open -> make change -> "Apply" -> "testing connection" spinner appears inside dialog -> "Connection will be lost" error inside dialog and buttons change to "keep connection", "apply anyway".

Which one do you prefer?

The first should be trivial to implement given that we need you mock up flow anyway for the direct buttons, so I'll start with that and we can see how well that works.

**Comment by mvollmer** [comment]
@andreasn, I have implemented your mockup with the full page curtain, both for direct actions on the page (like On/Off), and for changes via dialogs.

I think this works quite well, and having the same "Testing connection" -> "Connection will be lost" flow in all cases is probably the right thing.  Sorry for overcomplicating things, again. :)

**Comment by andreasn** [comment]
Cool. I'll test it out ASAP

**Comment by mvollmer** [comment]
Video: https://www.youtube.com/watch?v=fNIaDl5UR4Q

**Comment by andreasn** [comment]
Looks good to me!

**Comment by mvollmer** [comment]
One thing to keep in mind is that **every** change, including the harmless ones, now involves the "Testing connection" intermission.

**Comment by stefwalter** [comment]
> One thing to keep in mind is that every change, including the harmless ones, now involves the "Testing connection" intermission.

I was wondering about that. Can we put in a delay ... so that screen isn't shown unless testing takes more than a second?

**Comment by mvollmer** [comment]
> Can we put in a delay ... so that screen isn't shown unless testing takes more than a second?

The testing always takes the same time (2 seconds right now).  This is because we don't know how fast after a change the connection actually goes down, so we just wait 2 seconds and then try to destroy the checkpoint.  (If that succeeds, all is fine, if it fails, we assume that it has been rolled back.)

But, yes, I agree that we should try to reduce that time.  No delay definitely does not work, but 200ms might be as good as 2 seconds.

**Comment by andreasn** [comment]
Delay before throwing up the spinner sounds good to me.

**Comment by mvollmer** [comment]
> No delay definitely does not work, but 200ms might be as good as 2 seconds.

Yeah, can't say this is scientific, but 300ms seems to work.  Now the curtain only comes up when the connection really has broken.

**Comment by stefwalter** [comment]
Several related test failures:

```
Page error: #confirm-breaking-change-popup button:contains('Close') not found
phantomjs://code/phantom-lib.js 17 ph_only
phantomjs://code/phantom-lib.js 26 ph_find
phantomjs://code/phantom-lib.js 97 ph_click
 1 
 1 
```


```
    21: testBond (check_networking.TestNetworking)), duration: 153s screenshot journal
    23: testCheckpoint (check_networking.TestNetworking)), duration: 65s screenshot journal
    25: testTeam (check_networking.TestNetworking)), duration: 137s screenshot journal 
```

**Comment by mvollmer** [comment]
> Page error: #confirm-breaking-change-popup button:contains('Close') not found

Fixed.

> testBond (check_networking.TestNetworking)), duration: 153s screenshot journal

This is interesting.  (Cockpit rejected the bond creation because it would break the connection.) The only way I can explain this false rejection is that the rollback timeout of 5 seconds was too short: It took longer than 5 seconds to create the bond and NetworkManager lost its nerves and rolled back the checkpoint in the middle of it.

I have documented the whole procedure and increased the rollback timeout.

It would be best to start the "count down to rollback" only after the actual change has been done, but the API doesn't allow that.

> testTeam (check_networking.TestNetworking)), duration: 137s screenshot journal 

I think this is not related.  I can't reproduce it locally...

File: test/verify/check-networking [hunk 1]
```diff
@@ -420,5 +420,33 @@ NM_CONTROLLED=no
         self.browser.wait_present(sel)
         self.browser.wait_visible(sel)
 
+    @unittest.skipIf("fedora-25" not in os.environ.get("TEST_OS", ""), "No network checkpoints here.")
```
Comment (dperpeet) [inline test/verify/check-networking]: Do we have checkpoints everywhere else? What about fedora-23, fedora-24 and fedora-testing?

Comment (mvollmer) [inline test/verify/check-networking]: Checkpoints have been added to NetworkManager 1.4, which is only available in Fedora 25.

Comment (dperpeet) [inline test/verify/check-networking]: ok, you answered on irc: only on fedora-25 at the moment

**Pull request closed without merging**